src/rpc/request.cpp

```diff
@@ -85,7 +85,7 @@ bool GenerateAuthCookie(std::string *cookie_out)
      */
     fsbridge::ofstream file;
     fs::path filepath_tmp = GetAuthCookieFile(true);
-    file.open(filepath_tmp);
+    file.open(filepath_tmp.c_str());
```

Comment:
Is the `c_str()` actually needed? On my system (g++ 11.1) this also compiles when I write `file.open(filepath_tmp);`. This calls `operator string_type() const { return _M_pathname; }` from `std::fileystem::path`.

Or maybe add `operator string_type() const = delete;` to explicitly delete that conversion in the path wrapper

Reply:
c_str is needed on some platform (mingw, IIRC) because of an `_If_fs_path` requirement there which requires std::filesystem::path class argument or a string argument, and won't accept a std::filesystem::path subclass like fs::path (https://github.com/bitcoin/bitcoin/pull/20744#discussion_r733637362, https://github.com/bitcoin/bitcoin/pull/20744#issuecomment-948358014)

Reply:
Thanks!

Reply:
Wouldn't it be better to somehow do the c_str call hidden inside the fsbridge namespace? Seems a bit odd to force devs to write ugly and unintuitive code just because mingw can't even.

Reply:
> Wouldn't it be better to somehow do the c_str call hidden inside the fsbridge namespace? Seems a bit odd to force devs to write ugly and unintuitive code just because mingw can't even.

I also like the idea of hiding this in the fsbridge namespace (where there is no need to use c_str() at all), which is why I [originally suggested](https://github.com/bitcoin/bitcoin/pull/20744#discussion_r733637362) fixing this with:

```diff
-    typedef std::ofstream ofstream;
+    class ofstream : public std::ofstream
+    {
+        using std::ofstream::ofstream;
+        ofstream(const fs::path& p) : ofstream(static_cast<const std::filesystem::path&>(p)) {}
+    };
```

But @hebasto preferred the second alternative I suggested, which was calling to c_str() so we went with that approach instead. I still do like not calling c_str() better, but don't have a strong opinion.

Reply:
Sorry, if my words were misleading, but I expressed my preference for `c_str()` over a `static_cast`. I'm ok if it will be hidden in `fsbridge` namespace.

Reply:
> I still do like not calling c_str() better,

What reasons for?

Reply:
> > I still do like not calling c_str() better,
> 
> What reasons for?

Mainly just for the reason that marco wrote https://github.com/bitcoin/bitcoin/pull/20744#issuecomment-1022030719 that the `.c_str()` calls look ugly and and unsafe even if they are safe. It could be confusing that `path.c_str()` is safe when `path.string().c_str()` is unsafe, and it's nice to just be able to pass plain `path`. Also just passing the path object directly avoids temporary objects. No strong preference though, and no problem with either approach.

Reply:
Would it be useful to summarize what one should do and what shouldn't with the new filesystem API with regard to paths? Or is it somewhere? I mean I know ryanofsky wrote many useful comments in this PR and in others. Maybe we can have a few bullet points saying "this is OK, this is not" so that there are actually more people who can review PRs with FS changes.

Reply:
> Would it be useful to summarize what one should do and what shouldn't with the new filesystem API with regard to paths? Or is it somewhere? I mean I know ryanofsky wrote many useful comments in this PR and in others. Maybe we can have a few bullet points saying "this is OK, this is not" so that there are actually more people who can review PRs with FS changes. Just an idea. Feel free to ignore.

I think main recommendation is to use fs::path object wherever possible and avoid representing paths as strings. Beyond that there is a note in developer notes:

>- Use `fs::path::u8string()` and `fs::u8path()` functions when converting path
>  to JSON strings, not `fs::PathToString` and `fs::PathFromString`
>
>   - *Rationale*: JSON strings are Unicode strings, not byte strings, and
>      RFC8259 requires JSON to be encoded as UTF-8.

And a detailed comment on the [`PathToString`](https://github.com/bitcoin/bitcoin/blob/e3699b71c46bf66cfe363fc76ab8761dc39555a7/src/fs.h#L102-L120) function saying when it should be used.

You could write more documentation, but I think current documentation is at least not bad.